### ISSUE TITLE
User reported PicoBoard causing computer to crash

### ISSUE THREAD

**Issue by jwzimmer** [LLK/scratch-flash#1262]
**TL;DR:** A user reported that when trying to use a Picoboard with Scratch, "the Picoboard circle in Scratch goes green for a second, then yellow, the my computer crashes." After installing a windows driver, the boards are working more consistently, but there was another crash after that point.

Windows 7, Dell laptop, Chrome, Production

**====**
Reported via freshdesk no. 65598:

> I recently bought 4 SparkFun Picoboard Starter Kits through Amazon, and I haven't been able to get them to work on my school computers.

> I am running windows 7 on a Dell laptop. I use Chrome as my browser.

> This is what I did:
> 1. I went into Chrome and enabled the Scratch Picoboard extension
> 2. Then I downloaded the picoboard driver
> 3. Shut down computer
> 4. Plugged in Picoboard 
> 5. Turned on the computer (Computer told me that my device was installed properly on Com3)
> 6. Opened Google Chrome
> 7. Opened Scratch
> 8. Create a project in Scratch
> 9. Add an extension, Picoboard

> Then the Picoboard circle in Scratch goes green for a second, then yellow, the my computer crashes.

...

> The picoboards are running pretty well now. I had to install the windows driver. 

> The picoboard made one of my computers crash the other day, but working pretty well overall. No more messages about inappropriate content. 

I have asked the user which Windows driver they installed. I think it's the driver in step 2 of https://wiki.scratch.mit.edu/wiki/PicoBoard, but I figured checking couldn't hurt. I also asked if they can send us the logs from after any of the crashes.

**Comment by cpseager** [comment]
@jwzimmer This issue has been around for a long time for the picoboard plugin, we see it often in schools. The computer crash is not caused by Scratch itself, but by the device manager plugin trying to open the wrong COM port as it tries to connect to picoboard.

**Short version (workaround)**

1) change the allocated COM port (e.g. COM3) to COM1 (the school admin will need to do this, with the picoboard already plugged in):
Control Panel>Hardware>Device Manager>Ports (COM & LPT)>[select appropriate FTDI USB Serial Port & right click]>Properties>Port Settings>Advanced>COM Port Number

2) Make sure the picoboard is always connected to the USB cable, in the same USB port, before trying to connect. Always using the same USB socket on the computer is important, because a different virtual COM port number is allocated to each physical socket.

**Long Version**

See point 5 above for the issue (FTDI virtual COM port is set to COM3). The current way the plugin first tries to find a picoboard is to try to open ports in numerical order (e.g. COM1, COM2, COM3 etc) until it finds a port that it can open and then sends it ping data to see if it is a picoboard. This is quite naughty, what should happen is that Scratch should display a drop down list of available COM ports and ask you which one you want to open (so the user can select just COM3 then proceed) - and then Scratch should pass that COM port number directly across to the plugin so the plugin knows exactly which port to open.

There are all sorts of COM port drivers in use in schools that will crash if you try to open a port that is already being regularly scanned for activity (interactive whiteboard manufacturers  are by far the worst offenders, but there are also some others like dataloggers). If one of these drivers happens to be installed and already set to work on COM1 or COM2 you can then have problems with the picoboard set to COM3, because the plugin will try and open/ping COM1 and COM2 before it gets to COM3 and hence can inadvertently cause a system crash via the other driver.  In theory Windows should only allow one COM port to open in one application at a time, but there are lots of badly coded drivers out there that don’t respect this.

And what if the incorrect COM port opened (COM1 or COM2) is valid and already connected to a different hardware device altogether like a UPS? Who knows how that device will behave when it starts getting sent ping data from Scratch as part of the connection process…

**Comment by jwzimmer** [comment]
@cpseager Thank you very much for this info! It is super helpful. 👍
